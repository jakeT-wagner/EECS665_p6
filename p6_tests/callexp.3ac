[BEGIN GLOBALS]
[END GLOBALS]
[BEGIN fn_b LOCALS]
b (formal arg of 8)
[END fn_b LOCALS]
fun_fn_b:   enter fn_b
            getarg 1 [b]
lbl_0:      leave fn_b
[BEGIN fn_i LOCALS]
tmp0 (tmp var of 8 bytes)
[END fn_i LOCALS]
fun_fn_i:   enter fn_i
            [tmp0] := 2 ADD64 2
            setret [tmp0]
            goto lbl_1
lbl_1:      leave fn_i
[BEGIN fn_s LOCALS]
s (local var of 8 bytes)
[addrTmp0] (tmp loc of 8 bytes)
[addrTmp1] (tmp loc of 8 bytes)
[END fn_s LOCALS]
fun_fn_s:   enter fn_s
            [addrTmp0] := [s]
            RECEIVE [[addrTmp0]]
            [addrTmp1] := [s]
            setret [[addrTmp1]]
            goto lbl_2
lbl_2:      leave fn_s
[BEGIN v LOCALS]
[END v LOCALS]
fun_v:      enter v
lbl_3:      leave v
[BEGIN main LOCALS]
a (local var of 8 bytes)
b (local var of 8 bytes)
p (local var of 8 bytes)
tmp0 (tmp var of 8 bytes)
tmp1 (tmp var of 8 bytes)
tmp2 (tmp var of 8 bytes)
tmp3 (tmp var of 8 bytes)
tmp4 (tmp var of 8 bytes)
tmp5 (tmp var of 8 bytes)
tmp6 (tmp var of 8 bytes)
tmp7 (tmp var of 8 bytes)
tmp8 (tmp var of 8 bytes)
tmp9 (tmp var of 8 bytes)
tmp10 (tmp var of 8 bytes)
tmp11 (tmp var of 8 bytes)
tmp12 (tmp var of 8 bytes)
tmp14 (tmp var of 8 bytes)
[addrTmp13] (tmp loc of 8 bytes)
[END main LOCALS]
main:       enter main
            call v
            call fn_i
            call fn_i
            getret [tmp1]
            [a] := [tmp1]
            call fn_i
            getret [tmp2]
            [tmp3] := [tmp2] ADD64 [a]
            [a] := [tmp3]
            call fn_s
            getret [tmp4]
            [tmp5] := [tmp4]
            [tmp6] := 3
            call fn_i
            getret [tmp7]
            call fn_s
            getret [tmp8]
            [tmp9] := [tmp8]
            [tmp10] := [tmp7] DIV64 [tmp9]
            [tmp11] := [tmp6] MULT64 [tmp10]
            [tmp12] := [tmp5] SUB64 [tmp11]
            [addrTmp13] := [p]
            [tmp14] := [tmp12] ADD64 [[addrTmp13]]
            [b] := [tmp14]
            REPORT [a]
lbl_4:      leave main

