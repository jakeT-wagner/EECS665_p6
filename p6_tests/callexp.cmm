bool fn_b(bool b) {

}

int fn_i() {
    return 2 + 2;
}

short fn_s() {
    ptr short s;
    read @s;
    return @s;
}

void v(int a, int b, string c, short d, bool e, ptr string f, string g){
    return;
}

int main() {
    int a;
    string str;
    v(1, 3S, "abc", 4S, false, &str, str);

    fn_i();

    a = fn_i();

    a = fn_i() + a;

    int b; 
    ptr int p;

    b = fn_s() - 3S * (fn_i() / fn_s()) + @p;

    write a;
}