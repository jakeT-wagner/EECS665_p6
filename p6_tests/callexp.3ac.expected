[BEGIN GLOBALS]
str_0 "abc"
[END GLOBALS]
[BEGIN fn_b LOCALS]
b (formal arg of 8)
[END fn_b LOCALS]
fun_fn_b:   enter fn_b
            getarg 1 [b]
lbl_0:      leave fn_b
[BEGIN fn_i LOCALS]
tmp0 (tmp var of 8 bytes)
[END fn_i LOCALS]
fun_fn_i:   enter fn_i
            [tmp0] := 2 ADD64 2
            setret [tmp0]
            goto lbl_1
lbl_1:      leave fn_i
[BEGIN fn_s LOCALS]
s (local var of 8 bytes)
[addrTmp0] (tmp loc of 8 bytes)
[addrTmp1] (tmp loc of 8 bytes)
[END fn_s LOCALS]
fun_fn_s:   enter fn_s
            [addrTmp0] := [s]
            RECEIVE [[addrTmp0]]
            [addrTmp1] := [s]
            setret [[addrTmp1]]
            goto lbl_2
lbl_2:      leave fn_s
[BEGIN v LOCALS]
a (formal arg of 8)
b (formal arg of 8)
c (formal arg of 8)
d (formal arg of 8)
e (formal arg of 8)
f (formal arg of 8)
g (formal arg of 8)
[END v LOCALS]
fun_v:      enter v
            getarg 1 [a]
            getarg 2 [b]
            getarg 3 [c]
            getarg 4 [d]
            getarg 5 [e]
            getarg 6 [f]
            getarg 7 [g]
            goto lbl_3
lbl_3:      leave v
[BEGIN main LOCALS]
a (local var of 8 bytes)
str (local var of 8 bytes)
b (local var of 8 bytes)
p (local var of 8 bytes)
tmp0 (tmp var of 8 bytes)
tmp1 (tmp var of 8 bytes)
tmp2 (tmp var of 8 bytes)
tmp3 (tmp var of 8 bytes)
tmp4 (tmp var of 8 bytes)
tmp5 (tmp var of 8 bytes)
tmp6 (tmp var of 8 bytes)
tmp7 (tmp var of 8 bytes)
tmp8 (tmp var of 8 bytes)
tmp9 (tmp var of 8 bytes)
tmp10 (tmp var of 8 bytes)
tmp11 (tmp var of 8 bytes)
tmp12 (tmp var of 8 bytes)
tmp13 (tmp var of 8 bytes)
tmp14 (tmp var of 8 bytes)
tmp16 (tmp var of 8 bytes)
[addrTmp15] (tmp loc of 8 bytes)
[END main LOCALS]
main:       enter main
            [tmp0] := 3
            [tmp1] := str
            setarg 1 1
            setarg 2 [tmp0]
            setarg 3 str_0
            setarg 4 4
            setarg 5 0
            setarg 6 [tmp1]
            setarg 7 [str]
            call v
            call fn_i
            call fn_i
            getret [tmp3]
            [a] := [tmp3]
            call fn_i
            getret [tmp4]
            [tmp5] := [tmp4] ADD64 [a]
            [a] := [tmp5]
            call fn_s
            getret [tmp6]
            [tmp7] := [tmp6]
            [tmp8] := 3
            call fn_i
            getret [tmp9]
            call fn_s
            getret [tmp10]
            [tmp11] := [tmp10]
            [tmp12] := [tmp9] DIV64 [tmp11]
            [tmp13] := [tmp8] MULT64 [tmp12]
            [tmp14] := [tmp7] SUB64 [tmp13]
            [addrTmp15] := [p]
            [tmp16] := [tmp14] ADD64 [[addrTmp15]]
            [b] := [tmp16]
            REPORT [a]
lbl_4:      leave main